File: frontend/src/types/domain.ts
export type ElementType =
  | "non_utilise"
  | "accotement"
  | "trottoir"
  | "bau"
  | "cycle"
  | "voie"
  | "tpc";

export type Side = "gauche" | "droite";
export type StateKind = "existant" | "projet";
export type SourceMethod =
  | "entree_manuelle"
  | "menu_deroulant"
  | "recuperation_entrees"
  | "recuperation_dxf";
export type ComparisonStatus = "ameliore" | "inchange" | "degrade";

export const ELEMENT_TYPE_LABELS: Record<ElementType, string> = {
  non_utilise: "Non utilisé",
  accotement: "Accotement",
  trottoir: "Trottoir",
  bau: "BAU",
  cycle: "Cycle",
  voie: "Voie",
  tpc: "TPC",
};

export const SOURCE_LABELS: Record<SourceMethod, string> = {
  entree_manuelle: "Entrée manuelle",
  menu_deroulant: "Menu déroulant",
  recuperation_entrees: "Récupération entrées",
  recuperation_dxf: "Récupération automatique",
};

export const SOURCE_COLORS: Record<SourceMethod, string> = {
  entree_manuelle: "#ffd54f",
  menu_deroulant: "#4fc3f7",
  recuperation_entrees: "#81c784",
  recuperation_dxf: "#e0e0e0",
};

export interface Band {
  band_id: string;
  state: StateKind;
  side: Side;
  element_type: ElementType;
  source: SourceMethod;
  confidence: number;
  label_hint?: string | null;
  sample_count: number;
  width_min?: number | null;
  width_max?: number | null;
  width_mean?: number | null;
}

export interface Threshold {
  element_type: ElementType;
  reduit_m: number;
  standard_m: number;
}

export interface WidthSample {
  pk: number;
  side: Side;
  element_type: ElementType;
  state: StateKind;
  width_m?: number | null;
  source: SourceMethod;
  band_id?: string | null;
  note?: string | null;
  // Plan-view (true x, y) boundary points bounding this band at this PK, for
  // DXF plan reconstruction. Populated only when the extraction method has
  // real boundary geometry to draw from (DXF heuristic mode, IFC) — left null
  // for DXF calque/cote mode, which only ever has a (pk, value) text label,
  // no boundary line to reconstruct a plan curve from.
  near_x?: number | null;
  near_y?: number | null;
  far_x?: number | null;
  far_y?: number | null;
}

export const DEFAULT_THRESHOLDS: Threshold[] = [
  { element_type: "bau", reduit_m: 2.5, standard_m: 3.25 },
  { element_type: "voie", reduit_m: 7.5, standard_m: 8.0 },
  { element_type: "accotement", reduit_m: 1.0, standard_m: 2.5 },
  { element_type: "trottoir", reduit_m: 1.5, standard_m: 2.0 },
  { element_type: "cycle", reduit_m: 1.5, standard_m: 2.0 },
  { element_type: "tpc", reduit_m: 1.0, standard_m: 3.0 },
];

export const DEFAULT_DELTA_SEUIL_M = 0.05;

export interface RatioResult {
  side: Side;
  element_type: ElementType;
  state: StateKind;
  pct_sous_reduit: number;
  pct_entre: number;
  pct_sur_standard: number;
  n_samples: number;
}

export interface ComparisonRow {
  pk: number;
  side: Side;
  element_type: ElementType;
  width_existant?: number | null;
  width_projet?: number | null;
  delta?: number | null;
  status?: ComparisonStatus | null;
  // Plan-view (true x, y) boundary points at this row's PK, for drawing the
  // Comparatif DXF layer "en situation" like Ratios/Existant/Projet — every
  // row's pk exactly matches an original WidthSample's own pk (see
  // compareStates), so these are that sample's own near/far, not a new
  // interpolation.
  near_x?: number | null;
  near_y?: number | null;
  far_x?: number | null;
  far_y?: number | null;
}

export type UploadRole = "axes_profils" | "existant" | "projet";

File: backend/app/models/domain.py
"""Shared domain vocabulary used by every extraction, calculation and export module.

Everything downstream of extraction (ratios, comparison, Excel/DXF export) operates
only on these types, so DXF and IFC extractors are interchangeable: they both produce
a list of WidthSample and a list of Band, nothing else leaks through.
"""

from __future__ import annotations

from enum import Enum
from typing import Optional

from pydantic import BaseModel, Field


class ElementType(str, Enum):
    """Cross-section element categories, matching the 'Entrées' legend of the
    reference Excel workbook. Kept as an open, user-remappable vocabulary rather
    than hard-coded Route/Autoroute cross-section shapes."""

    NON_UTILISE = "non_utilise"
    ACCOTEMENT = "accotement"
    TROTTOIR = "trottoir"
    BAU = "bau"
    CYCLE = "cycle"
    VOIE = "voie"
    TPC = "tpc"


class Side(str, Enum):
    GAUCHE = "gauche"
    DROITE = "droite"


class StateKind(str, Enum):
    EXISTANT = "existant"
    PROJET = "projet"


class SourceMethod(str, Enum):
    """How a value was obtained. Mirrors the 4-level color legend in the reference
    workbook (Entrée manuelle / Menu déroulant / Récupération entrées / Récupération
    dxf) so the UI can render the same confidence badges. RECUPERATION_DXF is kept
    generic and also covers automatic extraction from an IFC source file."""

    ENTREE_MANUELLE = "entree_manuelle"
    MENU_DEROULANT = "menu_deroulant"
    RECUPERATION_ENTREES = "recuperation_entrees"
    RECUPERATION_DXF = "recuperation_dxf"

    @property
    def confidence(self) -> float:
        return {
            SourceMethod.RECUPERATION_ENTREES: 1.0,
            SourceMethod.MENU_DEROULANT: 1.0,
            SourceMethod.ENTREE_MANUELLE: 1.0,
            SourceMethod.RECUPERATION_DXF: 0.6,
        }[self]


class SourceFormat(str, Enum):
    DXF = "dxf"
    IFC = "ifc"


class ComparisonStatus(str, Enum):
    AMELIORE = "ameliore"
    INCHANGE = "inchange"
    DEGRADE = "degrade"


class WidthSample(BaseModel):
    """A single width measurement at one station, for one band, one state."""

    pk: float
    side: Side
    element_type: ElementType
    state: StateKind
    width_m: Optional[float] = None
    source: SourceMethod
    band_id: Optional[str] = None
    note: Optional[str] = None


class Band(BaseModel):
    """A strip between two adjacent cross-section boundary lines (DXF) or a single
    IfcPavement instance (IFC). This is the unit the manual-correction UI acts on:
    the automatic pass classifies the whole band at once (side + element_type),
    and a user override replaces that classification for every sample in the band
    without needing per-PK correction."""

    band_id: str
    state: StateKind
    side: Side
    element_type: ElementType
    source: SourceMethod
    confidence: float
    label_hint: Optional[str] = None  # layer name / IFC type name / heuristic note
    sample_count: int = 0
    width_min: Optional[float] = None
    width_max: Optional[float] = None
    width_mean: Optional[float] = None


class BandOverride(BaseModel):
    band_id: str
    side: Side
    element_type: ElementType


class Threshold(BaseModel):
    element_type: ElementType
    reduit_m: float
    standard_m: float


DEFAULT_THRESHOLDS: list[Threshold] = [
    Threshold(element_type=ElementType.BAU, reduit_m=2.50, standard_m=3.25),
    Threshold(element_type=ElementType.VOIE, reduit_m=7.50, standard_m=8.00),
    Threshold(element_type=ElementType.ACCOTEMENT, reduit_m=1.00, standard_m=2.50),
    Threshold(element_type=ElementType.TROTTOIR, reduit_m=1.50, standard_m=2.00),
    Threshold(element_type=ElementType.CYCLE, reduit_m=1.50, standard_m=2.00),
    Threshold(element_type=ElementType.TPC, reduit_m=1.00, standard_m=3.00),
]

DEFAULT_DELTA_SEUIL_M = 0.05


class RatioResult(BaseModel):
    side: Side
    element_type: ElementType
    state: StateKind
    pct_sous_reduit: float
    pct_entre: float
    pct_sur_standard: float
    n_samples: int


class ComparisonRow(BaseModel):
    pk: float
    side: Side
    element_type: ElementType
    width_existant: Optional[float] = None
    width_projet: Optional[float] = None
    delta: Optional[float] = None
    status: Optional[ComparisonStatus] = None
